syntax = "proto3";

import "google/protobuf/empty.proto";
import "google/protobuf/struct.proto";

package specter_proto;

// ----------------------------- PreviewerService ---------------------------- //

service PreviewerService {
    rpc ListenPreview (ObjectId) returns (stream PreviewImage) {}
}

// ------------------------------ MarkerService ------------------------------ //

service MarkerService {
    rpc Start (google.protobuf.Empty) returns (google.protobuf.Empty) {}
    rpc Stop (google.protobuf.Empty) returns (google.protobuf.Empty) {}
}

// ----------------------------- RecorderService ----------------------------- //

service RecorderService {
    rpc ListenCommands (google.protobuf.Empty) returns (stream RecorderCommand) {}
}

// ------------------------------ MouseService ------------------------------- //

service MouseService {
    rpc Click (MouseClick) returns (google.protobuf.Empty) {}
    rpc Move (MouseMove) returns (google.protobuf.Empty) {}
    rpc Scroll (MouseScroll) returns (google.protobuf.Empty) {}

    rpc ClickObject (ObjectMouseClick) returns (google.protobuf.Empty) {}
}

// ---------------------------- KeyboardService ------------------------------ //

service KeyboardService {
    rpc KeyPress (KeyEvent) returns (google.protobuf.Empty) {}
    rpc KeyRelease (KeyEvent) returns (google.protobuf.Empty) {}
    rpc TypeText (TextInput) returns (google.protobuf.Empty) {}

    rpc TypeIntoObject (ObjectTextInput) returns (google.protobuf.Empty) {}
}

// ----------------------------- ObjectService ------------------------------- //

service ObjectService {
    rpc GetTree (OptionalObjectId) returns (ObjectTree) {}
    rpc Find (ObjectSearchQuery) returns (ObjectIds) {}

    rpc GetObjectQuery (ObjectId) returns (ObjectSearchQuery) {}

    rpc GetParent (ObjectId) returns (ObjectId) {}
    rpc GetChildren (ObjectId) returns (ObjectIds) {}

    rpc CallMethod (MethodCall) returns (google.protobuf.Empty) {}
    rpc UpdateProperty (PropertyUpdate) returns (google.protobuf.Empty) {}

    rpc GetMethods (ObjectId) returns (Methods) {}
    rpc GetProperties (ObjectId) returns (Properties) {}

    rpc ListenTreeChanges (google.protobuf.Empty) returns (stream TreeChange) {}
    rpc ListenPropertiesChanges (ObjectId) returns (stream PropertyChange) {}
}

// -------------------------------- Messages --------------------------------- //

message ObjectId {
    string id = 1;
}

message OptionalObjectId {
    optional string id = 1;
}

message ObjectIds {
    repeated ObjectId ids = 1;
}

message ObjectSearchQuery {
    string query = 1;
}

message PreviewImage {
    bytes image = 1;
}

message RecorderCommand {
    string command = 1;
}

message ObjectTree {
    repeated ObjectNode roots = 1;
}

message ObjectNode {
    ObjectId object_id = 1;
    repeated ObjectNode children = 2;
}

message MethodCall {
    ObjectId object_id = 1;
    string method_name = 2;
    repeated google.protobuf.Value arguments = 3;
}

message PropertyUpdate {
    ObjectId object_id = 1;
    string property_name = 2;
    google.protobuf.Value value = 3;
}

message Methods {
    repeated Method methods = 1;
}

message Method {
    string method_name = 1;
    repeated Parameter parameters = 2;
}

message Parameter {
    string parameter_name = 1;
    google.protobuf.Value default_value = 2;
}

message Properties {
    repeated Property properties = 1;
}

message Property {
    string property_name = 1;
    google.protobuf.Value value = 2;
    bool read_only = 3;
}

message TreeChange {
    oneof change_type {
        ObjectAdded added = 1;
        ObjectRemoved removed = 2;
        ObjectReparented reparented = 3;
        ObjectRenamed renamed = 4;
    }
}

message ObjectAdded {
    ObjectId object_id  = 1;
    ObjectId parent_id  = 2;
}

message ObjectRemoved {
    ObjectId object_id  = 1;
}

message ObjectReparented {
    ObjectId object_id  = 1;
    ObjectId parent_id  = 2;
}

message ObjectRenamed {
    ObjectId object_id  = 1;
    ObjectSearchQuery object_query = 2;
}

message PropertyChange {
    oneof change_type {
        PropertyAdded added = 1;
        PropertyRemoved removed = 2;
        PropertyUpdated updated = 4;
    }
}

message PropertyAdded {
    string property_name = 1;
    google.protobuf.Value value = 2;
    bool read_only = 3;
}

message PropertyRemoved {
    string property_name = 1;
}

message PropertyUpdated {
    string property_name  = 1;
    google.protobuf.Value old_value = 2;
    google.protobuf.Value new_value = 3;
}

message MouseClick {
    int32 x = 1;
    int32 y = 2;
    enum Button {
        LEFT = 0;
        RIGHT = 1;
        MIDDLE = 2;
    }
    Button button = 3;
    bool double_click = 4;
}

message MouseMove {
    int32 x = 1;
    int32 y = 2;
}

message MouseScroll {
    int32 delta_x = 1;
    int32 delta_y = 2;
}

message ObjectMouseClick {
    ObjectId object_id = 1;
    MouseClick.Button button = 2;
    bool double_click = 3;
}

message KeyEvent {
    string key = 1; 
    bool ctrl = 2;
    bool alt = 3;
    bool shift = 4;
    bool meta = 5;
}

message TextInput {
    string text = 1;
}

message ObjectTextInput {
    ObjectId object_id = 1;
    string text = 2;
}